pkg/registry/ecr/registry.go

```diff
@@ -0,0 +1,118 @@
+package ecr
+
+import (
+	"encoding/json"
+	"os"
+
+	"github.com/v3io/registry-creds-handler/pkg/registry"
+	"github.com/v3io/registry-creds-handler/pkg/registry/abstract"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ecr"
+	"github.com/nuclio/errors"
+	"github.com/nuclio/logger"
+)
+
+type Registry struct {
+	*abstract.Registry
+	awsCreds registry.AWSCreds
+}
+
+func NewRegistry(parentLogger logger.Logger,
+	secretName string,
+	namespace string,
+	creds string,
+	registryUri string) (*Registry, error) {
+	newRegistry := &Registry{}
+
+	// create base
+	abstractRegistry, err := abstract.NewRegistry(parentLogger, newRegistry, secretName, namespace, creds, registryUri)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to create abstract registry")
+	}
+
+	newRegistry.Registry = abstractRegistry
+	return newRegistry, nil
+}
+
+func (r *Registry) EnrichAndValidate() error {
+	err := r.ValidateAbstractParams()
+	if err != nil {
+		return errors.Wrap(err, "Failed abstract registry params validation")
+	}
+
+	// parse aws credentials
+	var awsCreds registry.AWSCreds
+	if err := json.Unmarshal([]byte(r.Creds), &awsCreds); err != nil {
+		return errors.Wrap(err, "Failed to parse AWS credentials")
+	}
+
+	if awsCreds.Region == "" {
+		return errors.New("AWS Region is required")
+	}
+
+	if awsCreds.AccessKeyID == "" {
+		r.Logger.InfoWith("Did not receive AWS Access Key ID, checking env")
+		awsCreds.AccessKeyID = os.Getenv("AWS_ACCESS_KEY_ID")
+		if awsCreds.AccessKeyID == "" {
+			return errors.New("AWS Access Key ID is required")
+		}
+	}
+
+	if awsCreds.SecretAccessKey == "" {
+		r.Logger.InfoWith("Did not receive AWS Secret Access Key, checking env")
+		awsCreds.SecretAccessKey = os.Getenv("AWS_SECRET_ACCESS_KEY")
+		if awsCreds.SecretAccessKey == "" {
+			return errors.New("AWS Secret Access Key is required")
+		}
+	}
+
+	r.awsCreds = awsCreds
+
+	return nil
+}
+
+func (r *Registry) GetAuthToken() (*registry.Token, error) {
+	ecrClient := r.createECRClient()
+
+	r.Logger.DebugWith("Getting authorization token", "SecretName", r.SecretName, "Namespace", r.Namespace)
+	resp, err := ecrClient.GetAuthorizationToken(&ecr.GetAuthorizationTokenInput{})
+
+	if err != nil {
+		r.Logger.WarnWith("Failed to get ECR authorization token", "error", err.Error())
+		return nil, errors.Wrap(err, "Failed to get authorization token from ecr client")
+	}
+	r.Logger.DebugWith("Got GetAuthorizationToken response from ECR")
+
+	// This token has access to any ECR registry that the IAM principal has access to
+	for _, auth := range resp.AuthorizationData {
+		token := &registry.Token{
+			SecretName:  r.SecretName,
+			AccessToken: *auth.AuthorizationToken,
+			RegistryUri: r.RegistryUri,
+		}
+		r.Logger.InfoWith("Got authorization token", "ExpiresAt", auth.ExpiresAt)
+		return token, nil
+	}
```

Comment:
can you explain why is that a list? id expect it to be a single item response that is a token
what else can be returned within that list? meaning, what could possibly be a difference between item A and item B on that list

Reply:
`GetAuthorizationTokenInput` deprecated field:
```	
// Deprecated: This field is deprecated. The returned authorization token can be used to access any Amazon ECR registry that the IAM principal has access to, specifying a registry ID doesn't change the permissions scope of the authorization token.
RegistryIds []*string `locationName:"registryIds" min:"1" deprecated:"true" type:"list"`
```
so the corresponding `GetAuthorizationTokenOutput` field:
```
// A list of authorization token data objects that correspond to the registryIds
// values in the request.
AuthorizationData []*AuthorizationData `locationName:"authorizationData" type:"list"`
```

Basically, there is no reason for this to be a list after it's fully deprecated. Will add a comment to explain